app/views.py

```diff
@@ -223,3 +225,271 @@ def remove_single_item_from_cart(request, slug):
+            return redirect("app:order-summary")
+        
+@login_required
+@csrf_exempt
+def get_coupon(request, code):
+    try:
+        coupon = Coupon.objects.get(code=code)
+        return coupon
+    except ObjectDoesNotExist:
+        messages.info(request, _("This coupon does not exist"))
+        return redirect("app:checkout")
+
+class AddCouponView(LoginRequiredMixin, View):
+    def post(self, *args, **kwargs):
+        form = CouponForm(self.request.POST or None)
+        if form.is_valid():
+            try:
+                code = form.cleaned_data.get('code')
+                order = Order.objects.get(
+                    user=self.request.user, order_status=0)
+                order.coupon = get_coupon(self.request, code)
+                order.save()
+                messages.success(self.request, _("Successfully added coupon"))
+                return redirect("app:checkout")
+            except ObjectDoesNotExist:
+                messages.info(self.request, _("You do not have an active order"))
+                return redirect("app:checkout")
+
+@csrf_exempt
+def create_ref_code():
+    return ''.join(random.choices(string.ascii_lowercase + string.digits, k=20))
+
+class PaymentView(LoginRequiredMixin, View):
+    def get(self, *args, **kwargs):
+        order = Order.objects.get(user=self.request.user, order_status=0)
+        order_items = OrderItem.objects.filter(order=order)
+        coupon = 0
+        if (order.coupon == None):
+            coupon = 0
+        else:
+            coupon = order.coupon.amount
+        if order.billing_address:
+            context = {
+                'order': order,
+                'DISPLAY_COUPON_FORM': False,
+                'order_items' : order_items,
+                'total' : order.amount - coupon,
+            }
+            return render(self.request, "payment.html", context)
+        else:
+            messages.warning(
+                self.request, _("You have not added a billing address"))
+            return redirect("app:checkout")
+
+    def post(self, *args, **kwargs):
+        order = Order.objects.get(user=self.request.user, order_status=0)
+        form = PaymentForm(self.request.POST)
+        if form.is_valid():
+            card_number = form.cleaned_data.get('card_number')
+            coupon = 0
+            if (order.coupon == None):
+                coupon = 0
+            else:
+                coupon = order.coupon.amount
+
+
+            try:
+                # create the payment
+                payment = Payment()
+                payment.card_number = card_number
+                payment.user = self.request.user
+                payment.amount = order.amount - coupon
+                payment.status = 1
+                payment.save()
+
+                # assign the payment to the order
+                order.order_status = 1
+                order.payment = payment
+                order.ref_code = create_ref_code()
+                order.save()
```

Comment:
phần xử lý này có tác động đến 2 bảng trong DB -> e tìm hiểu dùng transaction để đảm bảo tính toàn vẹn và đúng đắn của dữ liệu nhé